### PR TITLE
Feature/enhance encryption

### PR DESCRIPTION
**Pull request by greymd**
This request includes following changes.

### New algorithm for stronger encryption (issue #29)
  + Please see https://github.com/greymd/ttcopy/blob/image/img/sequence.svg

### New exit codes

| Description                 | Exit status   | Note                                                              |
| -------------               | ------------- | ------                                                            |
| Failed to decode something. | 7             | i.e, Invalid password, ID                                         |
| Failed to encode something. | 8             | `openssl` got something wrong. i.e, `openssl`'s version is too old. |

After this request is merged, `master` branch will most likely be updated to be `v2.0.0`.

### CODE REVIEW

File: test/test.sh [hunk 1]
```diff
@@ -32,7 +32,13 @@ setUp () {
 # which is supposed to have 404 http status.
 cl1pMockserver () {
     local port=$1
-    printf "HTTP/1.0 200 Ok\n\nhttp://example.com/URL404/file.txt" | nc -l $port
+    # How to create it
+    # $ source ttcp.sh
```
Comment (mecab) [inline test/test.sh]: `lib/ttcp` ?

File: test/test.sh [hunk 2]
```diff
@@ -46,10 +52,37 @@ test_activator_dont_create_dupulicate_entry () {
 
     # As $_TTCP_DIR is already in the $PATH, so the activator should not do
     # anything. Here we check $PATH got no modification.
-
     assertEquals "$oldPATH" "$PATH"
 }
 
+test_failure_encoding () {
+    # openssl scommand stops with something wrong.
```
Comment (mecab) [inline test/test.sh]: typo? `scommand`

File: lib/ttcp [hunk 3]
```diff
@@ -45,10 +48,16 @@ _TTCP_DIR="$(cd "$(dirname "${BASH_SOURCE[0]:-${(%):-%N}}")"; cd ..; pwd)"
...
 
+# Algorithm for encoding data
+_TTCP_ENCRYPT_ALGORITHM="${_TTCP_ENCRYPT_ALGORITHM:-aes-256-cbc}"
+
 # Whare the last pasted content stored is.
 # It is re-used when you failed to get the remote content.
 TTCP_LASTPASTE_PATH_PREFIX="${TMPDIR}/lastPaste_"
+
 TTCP_ID_PREFIX="ttcopy"
+TTCP_SALT1="${TTCP_SALT1:-e7a082e7b396}"
```
Comment (mecab) [inline lib/ttcp]: Adding short explanation like is kind 😎 :
`Salt used to encrypt the url of the file you uploaded. You can change it to any value  (more secure).`

File: lib/ttcp [hunk 4]
```diff
@@ -45,10 +48,16 @@ _TTCP_DIR="$(cd "$(dirname "${BASH_SOURCE[0]:-${(%):-%N}}")"; cd ..; pwd)"
...
+TTCP_SALT2="${TTCP_SALT2:-e9a39fe5a1a9}"
```
Comment (mecab) [inline lib/ttcp]: Adding comment is kind again 😎

**Review by mecab — CHANGES_REQUESTED** [review]
Awesome! Better security! 🛡 👮 🛡 

I wrote a little reviews for documentation.

重箱の隅をつつくマンと化した。

File: .travis.yml [hunk 5]
```diff
@@ -13,6 +13,7 @@ before_install:
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]] ; then if [[ ! $(which $SH) ]] ; then sudo apt-get -qq update        ; fi ; fi
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]] ; then if [[ ! $(which $SH) ]] ; then sudo apt-get install -y $SH    ; fi ; fi
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]] ; then if [[ ! $(which nc) ]]  ; then sudo apt-get install -y netcat ; fi ; fi
+  - if [[ "$TRAVIS_OS_NAME" == "linux" ]] ; then if [[ ! $(which strings) ]]  ; then sudo apt-get install -y binutils ; fi ; fi # for `strings` command
```
Comment (mecab) [inline .travis.yml]: Where do you use `strings` command? I can't find by searching `strings`.

Comment (greymd) [reply .travis.yml]: `strings` command is used line 26 and 27 in `test.sh`.
This is used to generate random id/password.
Even though it is POSIX command, some of the OS does not have it (I tried it on the Ubuntu docker container, but it does not have the command...)

```bash
    export TTCP_ID="$(cat /dev/urandom | strings | grep -o '[[:alnum:]]' | tr -d '\n' | fold -w 128 | head -n 1)"
    export TTCP_PASSWORD="$(cat /dev/urandom | strings | grep -o '[[:alnum:]]' | tr -d '\n' | fold -w 128 | head -n 1)"
```

Anyway, I noticed that this part can be done without `strings` command like this.

```bash
cat /dev/urandom | grep -ao '[a-zA-Z0-9]' | tr -d '\n' | fold -w 128 | head -n 1
```

Therefore, I will delete the part you pointed out 👍

**Comment by greymd** [comment]
All the parts pointed out were fixed ☕️

**Comment by greymd** [comment]
Btw,

```sh
TTCP_SALT1="${TTCP_SALT1:-e7a082e7b396}"
TTCP_SALT2="${TTCP_SALT2:-e9a39fe5a1a9}"
```

```sh
$ echo e7a082e7b396 | xxd -r -p
砂糖

$ echo e9a39fe5a1a9 | xxd -r -p
食塩
```

（どうでもいい）

**Review by mecab — APPROVED** [review]

**Comment by mecab** [comment]
[LGTM](https://lgtm.in/i/C3sz1vwRp)

**Comment by mecab** [comment]
lol nice salt

**Pull request merged**